src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+                HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCONVERT, FAIL,
+                            "can't convert imaginary part of complex number");
+
+            /* If an exception was handled, go to the next element */
+            if (exception_handled) {
+                reverse = false;
+                goto next;
+            }
+        }
+
+padding:
+        /*
+         * Set external padding areas
+         */
+        if (dst_atomic.offset > 0) {
+            assert(H5T_PAD_ZERO == dst_atomic.lsb_pad || H5T_PAD_ONE == dst_atomic.lsb_pad);
+            H5T__bit_set(d, (size_t)0, dst_atomic.offset, (bool)(H5T_PAD_ONE == dst_atomic.lsb_pad));
+            H5T__bit_set(d + dst_part_size, (size_t)0, dst_atomic.offset,
+                         (bool)(H5T_PAD_ONE == dst_atomic.lsb_pad));
+        }
+        {
+            size_t type_size = dst_p->shared->parent->shared->size;
+
+            if (dst_atomic.offset + dst_atomic.prec != 8 * type_size) {
+                assert(H5T_PAD_ZERO == dst_atomic.msb_pad || H5T_PAD_ONE == dst_atomic.msb_pad);
+                H5T__bit_set(d, dst_atomic.offset + dst_atomic.prec,
+                             8 * type_size - (dst_atomic.offset + dst_atomic.prec),
+                             (bool)(H5T_PAD_ONE == dst_atomic.msb_pad));
+                H5T__bit_set(d + dst_part_size, dst_atomic.offset + dst_atomic.prec,
+                             8 * type_size - (dst_atomic.offset + dst_atomic.prec),
+                             (bool)(H5T_PAD_ONE == dst_atomic.msb_pad));
+            }
+        }
+
+        /* Put the destination in the correct byte order. See note at beginning of loop. */
+        if (H5T_ORDER_BE == dst_atomic.order && reverse) {
+            uint8_t *cur_part = d;
+            /* Swap real part of complex number element */
+            for (size_t j = 0; j < dst_part_size / 2; j++)
+                H5_SWAP_BYTES(cur_part, j, dst_part_size - (j + 1));
+            /* Swap imaginary part of complex number element */
+            cur_part += dst_part_size;
+            for (size_t j = 0; j < dst_part_size / 2; j++)
+                H5_SWAP_BYTES(cur_part, j, dst_part_size - (j + 1));
+        }
+        else if (H5T_ORDER_VAX == dst_atomic.order)
+            HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                        "VAX byte ordering is unsupported for complex number type conversions");
+
+next:
+        /*
+         * If we had used a temporary buffer for the destination then we
+         * should copy the value to the true destination buffer.
+         */
+        if (d == dbuf)
+            H5MM_memcpy(dp, d, dst_p->shared->size);
+
+        /* Advance source & destination pointers by delta amounts */
+        sp += src_delta;
+        dp += dst_delta;
+    } /* end conversion loop */
+
+done:
+    H5MM_free(src_rev);
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5T__conv_complex_loop() */
+
+/*-------------------------------------------------------------------------
```

Comment:
Maybe move this code to a 'static inline' routine in H5Tconv.h and invoke it from H5T__conv_f_f_loop and H5T__conv_complex_loop?   I bet that the compiler will inline it correctly and eliminate any performance gap.   Getting rid of the duplicated code would be very worthwhile from a maintenance perspective.

Reply:
Sure, I can probably do this. I must admit that I was a bit concerned about potential function call overhead in such a hot loop, but this function also isn't likely to be used very often unless applications make use of non-standard complex number types.